ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/PullQueryValidator.java

```diff
@@ -74,23 +92,84 @@ public void validate(final Analysis analysis) {
+        .stream()
+        .filter(col -> col instanceof SingleColumn) //filter out select *
+        .map(SingleColumn.class::cast)
+        .map(SingleColumn::getExpression)
+        .map(ColumnExtractor::extractColumns)
+        .flatMap(Collection::stream)
+        .map(ColumnReferenceExp::getColumnName)
+        .filter(name -> disallowedInPullQueries(name, analysis.getKsqlConfig()))
+        .map(Name::toString)
+        .collect(Collectors.joining(", "));
+
+    if (disallowedColumns.length() != 0) {
+      final String message = "Pull queries don't support the following columns in SELECT clauses: "
+          + disallowedColumns + "\n";
+      return Optional.of(message);
+    }
+
+    return Optional.empty();
+  }
+
+  private static Optional<String> disallowedColumnNameInWhereClause(final Analysis analysis) {
+    final Optional<Expression> expression = analysis.getWhereExpression();
+
+    if (!expression.isPresent()) {
+      return Optional.empty();
+    }
+
+    final String disallowedColumns = ColumnExtractor.extractColumns(expression.get())
+        .stream()
+        .map(ColumnReferenceExp::getColumnName)
+        .filter(name -> disallowedInPullQueries(name, analysis.getKsqlConfig()))
+        .map(Name::toString)
+        .collect(Collectors.joining(", "));
+
+    if (disallowedColumns.length() != 0) {
+      final String message = "Pull queries don't support the following columns in WHERE clauses: "
+          + disallowedColumns + "\n";
+      return Optional.of(message);
+    }
+
+    return Optional.empty();
+  }
+
+  private static boolean disallowedInPullQueries(
+      final ColumnName columnName,
+      final KsqlConfig ksqlConfig
+  ) {
+    final int pseudoColumnVersion = SystemColumns.getPseudoColumnVersionFromConfig(ksqlConfig);
+    return SystemColumns.isDisallowedInPullQueries(columnName, pseudoColumnVersion);
+  }
+
   private static final class Rule {
 
-    private final Predicate<Analysis> condition;
-    private final String failureMsg;
+    private final Function<Analysis, Optional<String>> potentialErrorMessageGenerator;
 
     private static Rule of(final Predicate<Analysis> condition, final String failureMsg) {
-      return new Rule(condition, failureMsg);
+      Function<Analysis, Optional<String>> potentialErrorMessageGenerator = (analysis) ->
+          !condition.test(analysis) ? Optional.of(failureMsg) : Optional.empty();
+
+      return new Rule(potentialErrorMessageGenerator);
     }
 
-    private Rule(final Predicate<Analysis> condition, final String failureMsg) {
-      this.condition = Objects.requireNonNull(condition, "condition");
-      this.failureMsg = Objects.requireNonNull(failureMsg, "failureMsg");
+    private static Rule of(final Function<Analysis, Optional<String>> function) {
+      return new Rule(function);
+    }
+
+    private Rule(final Function<Analysis, Optional<String>> function) {
+      this.potentialErrorMessageGenerator = Objects.requireNonNull(function);
```

Comment:
Use two-arg version of `requireNonNull()`

Reply:
Out of curiosity, why don't we have this as a checkstyle rule? Are there exceptions to the rule?

Reply:
I'm not sure I'd call this a style thing, more a code convention. It's not the end of the world if we slip up, but I think it's nice to be consistent throughout the codebase.